packages/@sanity/types/src/reference/types.ts

```diff
@@ -35,7 +35,8 @@ export type ReferenceFilterResolver = (options: {
 
 /** @public */
 export interface ReferenceFilterResolverOptions {
-  filter: ReferenceFilterResolver
+  filter?: ReferenceFilterResolver
+  filterParams?: never
```

Comment:
Had to add this never bit to allow TS to differentiate `ReferenceFilterResolverOptions` and  `ReferenceFilterQueryOptions`